File: 3085. Minimum Deletions to Make String K-Special/minimum_deletions_to_make_strings_k_special.py
class Solution:
    def minimumDeletions(self, word: str, k: int) -> int:
        def func(threshold: int) -> int:
            deletions = 0
            for frequency in frequencies:
                if frequency < threshold:
                    deletions += frequency
                # If frequency is greater than the threshold plus k, delete the excess
                elif frequency > threshold + k:
                    deletions += frequency - threshold - k
            return deletions
        frequencies = Counter(word).values()
        return min(func(threshold) for threshold in range(len(word) + 1))

  #T.C.->O(26^2)
  #S.C.->O(1)

File: 3085. Minimum Deletions to Make String K-Special/minimum_deletions_to_make_strings_k_special.cpp
class Solution {
public:
    int minimumDeletions(string word, int k) {
        vector<int> freq(26,0);
        for(char &ch:word)
        {
            freq[ch-'a']++;
        }
        int result=word.length();  //can be INT_MAX
        for(int i=0;i<26;i++)
        {
            int del=0;
            for(int j=0;j<26;j++)
            {
                if(i==j)
                    continue;
                if(freq[j]<freq[i])
                    del+=freq[j];
                else if(abs(freq[j]-freq[i])>k)
                    del+=abs(freq[j]-freq[i]-k);
            }
            result=min(result,del);
        }
        return result;
    }
};

// T.C.->O(26^2)=O(1)
// S.C.->O(1)
